manifests/charts/istio-control/istio-discovery/files/grpc-agent.yaml

```diff
@@ -1,6 +1,10 @@
 {{- $containers := list }}
 {{- range $index, $container := .Spec.Containers }}{{ if not (eq $container.Name "istio-proxy") }}{{ $containers = append $containers $container.Name }}{{end}}{{- end}}
 metadata:
+  labels:
+    security.istio.io/tlsMode: {{ index .ObjectMeta.Labels `security.istio.io/tlsMode` | default "istio"  | quote }}
```

Comment:
How will this be used? AFAIK auto mtls is not really compatible with proxyless gRPC since the application determines if it uses mTLS, not the control plane?

Reply:
Removed, thanks for the quick review.

To turn on mTLS, gRPC applications need to: 1. pass xDS credentials to client/server; 2. enable the mTLS in the control plane.

In the existing [mTLS setup guide](https://cloud.google.com/traffic-director/docs/security-proxyless-setup#configure-td-security) and (the [blog post](https://istio.io/latest/blog/2021/proxyless-grpc/#enabling-mtls)), I didn't find a pointer to this label (well, this is an Istio specific label). So, I removed it for now, as gRPC doesn't understand it.

In the long-term, could this flag help controlling the security feature for gRPC applications? Can I ask how does Envoy perform auto-mTLS?

CC @sanjaypujare

Reply:
In envoy, the configuration of the proxy is _100% controlled by XDS_. This is not the case in gRPC. 

Because we control the proxy 100%, the control plane is smart enough to know which servers at supporting mTLS and which are not. With gRPC we lose this.

I personally think it makes PSM a less effective product, although the reasons for doing it are reasonable. See discussion in https://github.com/grpc/grpc-go/issues/4123 and https://github.com/grpc/proposal/pull/184#discussion_r557617865.

This means that one of the largest friction points of a service mesh (configuring TLS properly) which has been solved in Istio already, will still occur on proxyless gRPC.

Reply:
Thanks for the context. I see now how gRPC applications need an extra step to enable mTLS.

Opt-in xDS at the library level is more invasive than opt-in the sidecar model. For gRPC clients, PSM requires the schema change, so they need to update anyway. For gRPC servers, you are right, this will be an extra step.

Reply:
> For gRPC clients, PSM requires the schema change, so they need to update anyway.

Strictly speaking the target string (which contains the schema `xds:`) can come from a run-time value such as a command line argument so strictly speaking even in the client it may not require code change. The specification of the channel credential (whether xDS creds or plaintext or TLS creds) can also depend on some run time value but more likely to be hardcoded in the code. The xDS-creds enables the app developer to control whether they want to rely on xDS provided security or use their own security. 

> For gRPC servers, you are right, this will be an extra step.

This is strictly speaking not true either. If they want an xDS-managed server they need to use the XdsServerBuilder (instead of the previous ServerBuilder) and the XdsServerBuilder then takes XdsServer creds or other types of creds but using XdsServer creds with fallback creds seems to be the standard option to use.